### PR TITLE
[SPARK-41204] [CONNECT] Migrate custom exceptions to use Spark exceptions

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
Migrate existing custom exceptions in Spark Connect to use the proper Spark exceptions.

### Why are the changes needed?
Consistency

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing UT

### CODE REVIEW

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -45,12 +46,18 @@ import org.apache.spark.util.Utils
 final case class InvalidPlanInput(
     private val message: String = "",
     private val cause: Throwable = None.orNull)
-    extends Exception(message, cause)
+    extends SparkException(
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:49]: cc @MaxGekk @itholic HELP!!

Comment (itholic) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:49]: Can we have test for the exception ??

If it's already exists, let's convert the existing test to use `checkError` to check error class and parameters generated properly.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -45,12 +46,18 @@ import org.apache.spark.util.Utils
...
+      errorClass = "CONNECT.INVALID_PLAN_INPUT",
+      messageParameters = Map("msg" -> message),
+      cause = cause)
 
 final case class InvalidCommandInput(
     private val message: String = "",
     private val cause: Throwable = null)
-    extends Exception(message, cause)
+    extends SparkException(
```
Comment (itholic) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:57]: ditto

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:59]: Done

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:51]: done

**Comment by grundprinzip** [comment]
@itholic Finally got around addressing your comments, please have another look. Thanks!

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala [hunk 3]
```diff
@@ -329,19 +346,25 @@ class SparkConnectPlannerSuite extends SparkFunSuite with SparkConnectPlanTest {
...
-      transform(proto.Relation.newBuilder.setDeduplicate(deduplicate).build())
-    }
-    assert(
-      e.getMessage.contains("Cannot deduplicate on both all columns and a subset of columns"))
+    checkError(
+      exception = intercept[InvalidPlanInput] {
+        transform(proto.Relation.newBuilder.setDeduplicate(deduplicate).build())
+      },
+      errorClass = "CONNECT.INVALID_PLAN_INPUT",
+      parameters = Map("msg" -> "Cannot deduplicate on both all columns and a subset of columns"))
```
Comment (MaxGekk) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:354]: Would it possible to move all the plain text to `error-classes.json`. Could you image that we will translate errors to local languages in the future, or tech editors will fix some typos in the texts. In your case, we will have to rebuild Spark which is not nice approach.

Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:354]: Happy to move the literals out.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -265,7 +280,10 @@ class SparkConnectPlanner(session: SparkSession) {
         // so we call filter instead of find.
         val cols = allColumns.filter(col => resolver(col.name, colName))
         if (cols.isEmpty) {
-          throw InvalidPlanInput(s"Invalid deduplicate column ${colName}")
+          throw new SparkException(
+            errorClass = "CONNECT.INVALID_PLAN_DEDUPLICATE_INVALID_INPUT_COLUMN",
+            messageParameters = Map("col" -> colName),
```
Comment (MaxGekk) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:285]: Could you quote the column by `toSQLId()`, please.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 5]
```diff
@@ -504,7 +545,11 @@ class SparkConnectPlanner(session: SparkSession) {
       case proto.Join.JoinType.JOIN_TYPE_LEFT_OUTER => LeftOuter
       case proto.Join.JoinType.JOIN_TYPE_RIGHT_OUTER => RightOuter
       case proto.Join.JoinType.JOIN_TYPE_LEFT_SEMI => LeftSemi
-      case _ => throw InvalidPlanInput(s"Join type ${t} is not supported")
+      case _ =>
+        throw new SparkException(
+          errorClass = "CONNECT.INVALID_PLAN_UNSUPPORTED_JOIN_TYPE",
+          messageParameters = Map("type" -> s"${t}"),
```
Comment (MaxGekk) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:551]: nit:
```suggestion
          messageParameters = Map("type" -> s"$t"),
```

File: core/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -134,6 +134,97 @@
...
+        "message" : [
+          "BucketBy must specify a bucket count > 0, received <bucket> instead."
+        ]
+      },
+      "INVALID_PLAN_INPUT" : {
+        "message" : [
+          "Error transforming Spark Connect plan: <msg>"
+        ]
+      },
+      "INVALID_PLAN_UNKNOWN_RELATION": {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:148]: Wrong the formatting here and bellow:
```suggestion
      "INVALID_PLAN_UNKNOWN_RELATION" : {
```

File: core/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -134,6 +134,97 @@
...
+        "message": [
+          "Relation: <param> not supported."
+        ]
+      },
+      "INVALID_PLAN_TRANSFORM_NA_FILL_EMPTY": {
+        "message": [
+          "Values must contains at least 1 item"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:155]:
```suggestion
          "Values must contains at least 1 item."
```

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -134,6 +134,97 @@
...
+        ]
+      },
+      "INVALID_PLAN_TRANSFORM_NA_FILL_SIZE": {
+        "message": [
+          "When values contains more than 1 items, ",
+          "values and cols should have the same length!"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:161]:
```suggestion
          "When values contains more than 1 items, values and cols should have the same length."
```

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -134,6 +134,97 @@
...
+        ]
+      },
+      "INVALID_PLAN_DEDUPLICATE_INVALID_INPUT_COLUMN" : {
+        "message": [
+          "Invalid column name used for Deduplicate: <col>"
+        ]
+      },
+      "INVALID_PLAN_INVALID_DATA_SOURCE" : {
+        "message": [
+          "Read from DataSource requires a format parameter"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:186]:
```suggestion
          "Read from DataSource requires a format parameter."
```

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -134,6 +134,97 @@
...
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_JOIN_TYPE" : {
+        "message": [
+          "Unsupported Join type: <type>"
+        ]
+      },
+      "INVALID_PLAN_INVALID_JOIN_CONDITION" : {
+        "message": [
+          "Using columns or join conditions cannot be set at the same time in Join"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:221]:
```suggestion
          "Using columns or join conditions cannot be set at the same time in Join."
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 5 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:551]: Done

File: core/src/main/resources/error/error-classes.json (hunk 6 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:148]: I regenerated the json file based on the SparkThrowableSuite.

File: core/src/main/resources/error/error-classes.json (hunk 9 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:186]: Done.

File: core/src/main/resources/error/error-classes.json (hunk 10 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:221]: Done.

File: core/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -132,7 +132,97 @@
...
+        "message" : [
+          "Using columns or join conditions cannot be set at the same time in Join."
+        ]
+      },
+      "INVALID_PLAN_INVALID_SET_OPERATION" : {
+        "message" : [
+          "Invalid Set operation: <type> does not support <param>."
+        ]
+      },
+      "INVALID_PLAN_TRANSFORM_NA_FILL_EMPTY" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:183]: Is it tested somewhere? If not, could add a test, please.

File: core/src/main/resources/error/error-classes.json [hunk 12]
```diff
@@ -132,7 +132,97 @@
...
+        "message" : [
+          "Values must contains at least 1 item."
+        ]
+      },
+      "INVALID_PLAN_TRANSFORM_NA_FILL_SIZE" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:188]: Is there a test for the error?

File: core/src/main/resources/error/error-classes.json [hunk 13]
```diff
@@ -132,7 +132,97 @@
...
+        "message" : [
+          "Relation: <param> not supported."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_DATA_SOURCE" : {
+        "message" : [
+          "Unsupported input data source: <type>."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_EXPRESSION_TYPE" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:203]: Could you add a test for this, please

File: core/src/main/resources/error/error-classes.json [hunk 14]
```diff
@@ -132,7 +132,97 @@
...
+        "message" : [
+          "Unsupported expression type: <type>."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_JOIN_TYPE" : {
+        "message" : [
+          "Unsupported Join type: <type>."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_LITERAL_TYPE" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:213]: Please, add a test for this error class.

File: core/src/main/resources/error/error-classes.json [hunk 15]
```diff
@@ -132,7 +132,97 @@
...
+        "message" : [
+          "Unsupported Literal type: <type>."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_SET_OPERATION" : {
+        "message" : [
+          "Unsupported Set operation: <type>."
+        ]
+      },
+      "INVALID_PLAN_UNSUPPORTED_TYPE_CONVERSION" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:223]: Check this error class, please.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
also cc @srielau

File: core/src/main/resources/error/error-classes.json (hunk 11 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:183]: Yes, there tests for all of these messages. Please see the SparkConnectPlannerSuite and SparkConnectProtoSuite.

**Comment by grundprinzip** [comment]
I will add tests for the missing cases.

File: core/src/main/resources/error/error-classes.json (hunk 11 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:183]: Done

File: core/src/main/resources/error/error-classes.json (hunk 12 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:188]: Done

File: core/src/main/resources/error/error-classes.json (hunk 15 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:223]: Done

File: core/src/main/resources/error/error-classes.json (hunk 14 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:213]: Done

File: core/src/main/resources/error/error-classes.json (hunk 13 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:203]: Done

**Review by grundprinzip — COMMENTED** [review]
Addressed the comments, added tests.

File: core/src/main/resources/error/error-classes.json (hunk 11 above)
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:183]: What is NA?

**Comment by srielau** [comment]
General comment. You use CONNECT as error class, everything else is a sub error class. Many of these are INVALID_PLAN.
How many errors total do you expect?
Note that at present we support only two levels.
Would it make sense to use CONNECT (or an abbreviation of it) as a prefix and then perhaps subclass at INVALID_PLAN:
CONNECT_INVALID_PLAN.xxx

File: core/src/main/resources/error/error-classes.json (hunk 11 above)
Comment (grundprinzip) [reply core/src/main/resources/error/error-classes.json:183]: It's the underlying operator.

```
 def fillna(
        self,
        value: Union["LiteralType", Dict[str, "LiteralType"]],
        subset: Optional[Union[str, Tuple[str, ...], List[str]]] = None,
    ) -> "DataFrame":
        """Replace null values, alias for ``na.fill()``.
        :func:`DataFrame.fillna` and :func:`DataFrameNaFunctions.fill` are aliases of each other.

```

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:183]: Uhm, I'm now no smarter. The real question is: Will users know what to do with this error?

**Pull request closed without merging**